app/server/app/utilities/streamingService.js

```diff
@@ -0,0 +1,141 @@
+        console.error({ error: Object.assign({}, error, { stack, message }) });
+      },
+    };
+  };
+
+  static writeHead = (res, status, format) => {
+    if (typeof res.headersSent === "boolean" && !res.headersSent) {
+      let contentType = "application/json; charset=utf-8";
+      if (format === "csv") contentType = "text/csv";
+      if (format === "tsv") contentType = "text/tsv";
+
+      res.writeHead(status, {
+        "Content-Type": contentType,
+        "Transfer-Encoding": "chunked",
+        "X-Content-Type-Options": "nosniff",
+      });
+    }
+  };
+
+  static getBasicTransform = (preHook, format) => {
+    return new Transform({
+      writableObjectMode: true,
+      transform(data, encoding, callback) {
+        // preHook on first data only
+        if (!this.comma) preHook();
+
+        // convert the json to csv
+        const unparsedData = Papa.unparse(JSON.stringify([data]), {
+          delimiter: format === "tsv" ? "\t" : ",",
+          header: !this.comma,
+        });
+        this.push(unparsedData);
+
+        // set comma for subsequent data
+        if (!this.comma) this.comma = "\n";
+        this.push(this.comma);
+        callback();
+      },
+      final(callback) {
+        callback();
+      },
+    });
+  };
+
+  static getJsonTransform = (preHook) => {
+    return new Transform({
+      writableObjectMode: true,
+      transform(data, encoding, callback) {
+        // preHook on first data only
+        if (!this.comma) preHook();
+        // if first data && error then no open/close brackets
+        const prefix = this.comma || (data.error ? "" : "[");
+        const suffix = this.comma && data.error ? "]" : "";
+        this.push(`${prefix}${JSON.stringify(data)}${suffix}`);
+        // set comma for subsequent data
+        if (!this.comma) this.comma = ",\n";
+        callback();
+      },
+      final(callback) {
+        if (!this.comma) this.push("[");
+        this.push("]");
+        callback();
+      },
+    });
+  };
+
+  static streamResponse = (outStream, inStream, format) => {
+    const { preHook, errorHook, errorHandler } = StreamingService.getOptions(
+      outStream,
+      format
+    );
+    inStream.on("error", (error) => {
+      errorHook();
+      inStream.push({ error: error.message });
+      outStream.end();
+    });
+
+    let transform = StreamingService.getJsonTransform(preHook);
+    if (format === "csv" || format === "tsv") {
+      transform = StreamingService.getBasicTransform(preHook, format);
+    }
+
+    pipeline(inStream, transform, outStream, errorHandler);
+  };
+}
+
+const ExcelTransform = function (options) {
+  Transform.call(this, {
+    writableObjectMode: true,
+    readableObjectMode: false,
+  });
+
+  this.workbook = options.workbook;
+  const that = this;
+  this.workbook.stream.on("readable", function () {
+    const chunk = this.workbook.stream.read();
+    that.push(chunk);
+  });
+  this.worksheet = options.worksheet;
+};
```

Comment:
Is the `that` necessary? It doesn't seem to make a difference in the result set for me.

Reply:
`that` has already been removed.